### DISCUSSION TITLE
Complex objective functions

### DISCUSSION THREAD

**Discussion by FedeClaudi** [infiniteopt/InfiniteOpt.jl#277] [Q&A] [answered]
Hi, 

I'm trying to implement an `@objective` function which requires calling `argmin` on a variable. The full story behind is a bit long, but basically I'm using a Bezier function to estimate the penalty cost for a term `Fu` and I need `argmin` to go from `Fu` to the parameter for the Bezier curve.

```julia
    # bezier curve
    t = 0:.001:1

    P0_x = -750
    P1_x = -350
    P2_x = -350
    P3_x = 3000
    
    P0_y = 1
    P1_y = 0
    P2_y = 0
    P3_y = 1

    Px(t) = P0_x * (1 - t)^3 + P1_x * 3(1 - t)^2 * t + P2_x * 3(1 - t) * t^2 + P3_x * t^3
    Py(t) = P0_y * (1 - t)^3 + P1_y * 3(1 - t)^2 * t + P2_y * 3(1 - t) * t^2 + P3_y * t^3
    X = Px.(t)

    cost(Fu) = Py(
        argmin(abs.(X .- Fu))
    )

    @objective(model, Min, ∫(SF + α*cost(Fu), s))
    optimize!(model)
```

However I then get this error:
```
MethodError: no method matching isless(::InfiniteOpt.NLPExpr, ::InfiniteOpt.NLPExpr)
```

I've tried changing it to `value(Fu)` but then I get an error saying that `Fu(s)` is not in the transcription model.

**Comment by pulsipher** [comment, marked as answer]
The problem is that `argmin` is not one of the accepted nonlinear functions. You can always check with `all_registered_functions` to see what functions can be used to build nonlinear expressions. Like JuMP, we can register new nonlinear functions to use. A current limitation of this interface is that each function must be scalar valued (not vector inputs/outputs). We hope to remove this limitation in the relatively near future (#229). 

In this case, we cannot register directly since `argmin` is vector valued. However, if `Fu` is a scalar (or we can easily enumerate it with scalar components) then we can register the cost function:
```julia
@register(model, cost(Fu))
```
And then use it when defining the objective.

**Reply by pulsipher** [reply]
For more info, see https://infiniteopt.github.io/InfiniteOpt.jl/stable/guide/expression/#Function-Registration

**Comment by FedeClaudi** [comment]
it worked perfectly, thank you!
